### PR TITLE
Customer return

### PR DESCRIPTION
**Pull request by aleph1ow**
Posts a return to Shipwire after saving a CustomerReturn.
Supports returning multiple line items with quantity.

### CODE REVIEW

File: app/models/customer_return_decorator.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
+module SolidusShipwire::CustomerReturn
```
Comment (DanielePalombo) [inline app/models/customer_return_decorator.rb]: Can you nest these modules?

File: app/models/customer_return_decorator.rb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+  def self.prepended(base)
+    base.after_save :process_shipwire_return!, if: :create_on_shipwire?
+  end
+
+  def to_shipwire
+    {
+      externalId: order.number,
+      originalOrder: {
+        id: order.shipwire_id.to_s
```
Comment (DanielePalombo) [inline app/models/customer_return_decorator.rb]: `to_s` is required here?

File: app/models/customer_return_decorator.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  end
+
+  private
+
+  def create_on_shipwire?
+    shipwire_order.present?
+  end
+
+  def process_shipwire_return!
+  create_on_shipwire
```
Comment (DanielePalombo) [inline app/models/customer_return_decorator.rb]: nest this please

File: app/models/customer_return_decorator.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  rescue SolidusShipwire::ResponseException => e
+    if e.response.validation_errors
```
Comment (DanielePalombo) [inline app/models/customer_return_decorator.rb]: if this condition is false we don't do anything?

File: app/models/customer_return_decorator.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+      errors.add(:shipwire, "Shipwire: #{e.response.validation_errors.first['message']}")
+    end
+  end
+
+  def to_shipwire_object(hash)
+    SolidusShipwire::ShipwireObjects::Return.new(hash['id'], self, hash)
+  end
+
+  def shipwire_line_items
```
Comment (DanielePalombo) [inline app/models/customer_return_decorator.rb]: This method `shipwire_line_items` is related to `return_items`, maybe can be better rename it in `shipwire_return_items`?

File: spec/models/spree/customer_return_spec.rb [hunk 6]
```diff
@@ -0,0 +1,128 @@
+describe Spree::CustomerReturn, type: :model do
+
+  # => Actually, just in case you want to regenerate vcr you must manually
```
Comment (DanielePalombo) [inline spec/models/spree/customer_return_spec.rb]: This logic is very complex! Do you think can be created an helper to do these steps and avoid human errors?

**Review by DanielePalombo — COMMENTED** [review]
great PR , I created only some note.

File: spec/models/spree/customer_return_spec.rb (hunk 6 above)
Comment (aleph1ow) [reply spec/models/spree/customer_return_spec.rb]: There is a TODO annotation that describes how to generate orders in shipwire with particular state via shipwire API. Actually I don't know if this is possible but I've left the instructions for a future implementation. I could move these notes to a new issue. What do you think about it @DanielePalombo ?

File: spec/models/spree/customer_return_spec.rb [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+  # TODO: Dynamic order creation and state update via Shipwire API
+  # => For the purpose of these tests we should dinamically create 4 orders
+  # => 1 -  Create a complete order that contains 1 line item with variant with sku_1
+  # => 2 -  Create a complete order that contains 1 line item with variant with sku_1, quantity: 1
+  # =>      and after save mark it as Shipped on shipwire (via api)
+  # => 4 -  Create a complete order that contains 1 line item with variant with sku_1, quantity: 2,
+  # =>      1 line item with variant with sku_2, quantity: 2
+  # =>      and after save mark it as Shipped on shipwire (via api)
+  # => 3 -  Create a complete order that contains 1 line item with variant with sku_1
```
Comment (kennyadsl) [inline spec/models/spree/customer_return_spec.rb]: why 3 after 4 ?

File: spec/models/spree/customer_return_spec.rb [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        let(:return_item)     { build(:return_item, inventory_unit: inventory_unit) }
+        let(:shipwire_id)     { pending_order_shipwire_id }
+        let(:return_items)    { [return_item] }
+
+        before do
+          customer_return.return_items = return_items
+          customer_return.order.update_attribute(:shipwire_id, shipwire_id)
+        end
+
+        it 'will post to shipwire' do
```
Comment (kennyadsl) [inline spec/models/spree/customer_return_spec.rb]: please use `posts` instead of `will post`

File: app/models/customer_return_decorator.rb (hunk 4 above)
Comment (aleph1ow) [reply app/models/customer_return_decorator.rb]: If the condition is false I do nothing. This is the case of a rare Timeout exception. All the other exceptions have validation errors and I forward them to the object. Timeout does not have any validation errors attached. How should I rescue from it?

File: spec/models/spree/customer_return_spec.rb [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+          expect(customer_return).to receive(:process_shipwire_return!).and_call_original
+          customer_return.save!
+        end
+
+        it 'add an error message' do
+          customer_return.save!
+          expect(customer_return.errors.messages).to have_key(:shipwire)
```
Comment (kennyadsl) [inline spec/models/spree/customer_return_spec.rb]: do error messages have important information? can we also test their content?

File: spec/models/spree/customer_return_spec.rb [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        end
+      end
+
+
+      context 'when a return for the order was already been reported.)',
```
Comment (kennyadsl) [inline spec/models/spree/customer_return_spec.rb]: there's an extra `)` at the end of this line

File: spec/models/spree/customer_return_spec.rb [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+      context 'when return has multiple return items',
+              vcr: { cassette_name: 'spree/customer_return_create_entity_multiple_return_items' } do
+
+        let(:customer_return) { build(:customer_return) }
+        let(:inventory_unit_1) { build(:inventory_unit, variant: variant_1) }
+        let(:inventory_unit_2) { build(:inventory_unit, variant: variant_2, order: inventory_unit_1.order) }
+        let(:return_item_1) { build(:return_item, inventory_unit: inventory_unit_1) }
+        let(:return_item_2) { build(:return_item, inventory_unit: inventory_unit_2) }
+        let(:shipwire_id)   { shipped_order_shipwire_id_2 }
+        let(:return_items)  { [return_item_1, return_item_1, return_item_2, return_item_2] }
```
Comment (kennyadsl) [inline spec/models/spree/customer_return_spec.rb]: can't you move these shared `let` outside the context so we avoid duplication? Or there's a reason to keep them duplicated?

File: spec/models/spree/customer_return_spec.rb [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+        before do
+          customer_return.return_items = return_items
+          customer_return.order.update_attribute(:shipwire_id, shipwire_id)
+        end
+
+        it 'will post to shipwire' do
+          expect(customer_return).to receive(:process_shipwire_return!).and_call_original
+          customer_return.save!
+          expect(customer_return.shipwire_id).not_to be nil
```
Comment (kennyadsl) [inline spec/models/spree/customer_return_spec.rb]: why in the previous content you used 2 scenarios and here just one? what's the difference?

File: app/controllers/spree/admin/customer_returns_controller_decorator.rb [hunk 13]
```diff
@@ -0,0 +1,17 @@
+module SolidusShipwire
+  module ShipwireCustomerReturnsController
+    def self.prepended(base)
+      base.after_action :shipwire, only: [:create]
+    end
+
+    private
+
+    def shipwire
```
Comment (kennyadsl) [inline app/controllers/spree/admin/customer_returns_controller_decorator.rb]: can't we find a better name for this method?

**Review by kennyadsl — DISMISSED** [review]

**Review by mtylty — APPROVED** [review]

**Pull request merged**